src/hooks/useSigningKeys.ts

```diff
@@ -0,0 +1,86 @@
+import { ethers } from "ethers";
+import { useState, useEffect, useContext } from "react";
+import { ErrorContext } from "common/context/ErrorContext";
+
+import { recoverPublicKey } from "features/vote/helpers/recoverPublicKey";
+import { derivePrivateKey } from "features/vote/helpers/derivePrivateKey";
+import usePrevious from "common/hooks/usePrevious";
+
+// User has to sign once per round.
+// public and private signing keys is indexed by address, and we ask them to resign if the roundMessage changes.
+export interface SigningKeys {
+  [address: string]: {
+    publicKey: string;
+    privateKey: string;
+    roundMessage: string;
+  };
+}
+
+export default function useSigningKeys(
+  signer: ethers.Signer | null,
+  address: string | null,
+  roundId: string
+) {
+  const [signingKeys, setSigningKeys] = useState<SigningKeys>(() => {
+    const storageKeys = localStorage.getItem("signingKeys");
+    return storageKeys ? (JSON.parse(storageKeys) as SigningKeys) : {};
+  });
+
+  const { addError } = useContext(ErrorContext);
+
+  const previousSigner = usePrevious(signer);
+  const previousAddress = usePrevious(address);
+
+  useEffect(() => {
+    if (
+      signer &&
+      address &&
+      roundId &&
+      (previousAddress === null || previousSigner === null)
+    ) {
+      const message = `UMA Protocol one time key for round: ${roundId}`;
```

Comment:
Is there any reason we can't just drop the round id from the message so people _never_ have to sign again? Or do we just want to slow roll this in case there are any issues?

Reply:
Yeah I think it's largely due to being a bit conservative about the signing process.

Nothing technically. We'd just want to have a good system in place to allow the user to *force* a re-sign if they needed to, or invalidate it if we know its wrong.